### DISCUSSION TITLE
Unable to build onnxruntime 1.6.0 from source

### DISCUSSION THREAD

**Discussion by shuiyu486** [microsoft/onnxruntime#7382] [Other Q&A] [answered]
**Describe the bug**
Hello there! I am a newbie and want to try to build onnxruntime, but I have encountered a problem and hope to help me, thank you.
Questions are as follows:
I downloaded the onnxruntime1.6.0 source code, and when I used vs2019 Developer Command Prompt for the simplest build, it failed.
![image](https://user-images.githubusercontent.com/38516221/115015661-97b7bd80-9ee6-11eb-9724-e80662e26880.png)

**System information**
- OS Platform :windows 10 1809
- ONNX Runtime version:1.6.0
- Python version:3.9.4(Tried python3.7.0 with the same failure)
- Visual Studio version :vs2019
- CMake :3.20.1
- CPU model: Intel Core i5-8400 @ 2.80GHz

**Comment by hariharans29** [comment]
Did you pull in all submodules ?

**Comment by shuiyu486** [comment]
> Did you pull in all submodules ?
> 
> 你把所有子模块都调出来了吗？

Thank you for replying to my help, do I need any sub-modules for the CPU version of windows? I did not add any sub-modules, just run the official bat script in the source folder.

**Comment by hariharans29** [comment]
I think that you are missing dependencies and yes you need dependencies even if building CPU version for Windows. Try running:
git submodule update --init --recursive and then re-try building.

**Comment by shuiyu486** [comment, marked as answer]
> I think that you are missing dependencies and yes you need dependencies even if building CPU version for Windows. Try running:
> git submodule update --init --recursive and then re-try building.


> 
Thanks, I can continue to build after using "git submodule update --init --recursive", but encountered a new failure (about pthread):：
![1](https://user-images.githubusercontent.com/38516221/115223653-a00a3580-a13e-11eb-9822-d53d673fce3a.PNG)



Check CMakeError.log:
`Performing C SOURCE FILE Test CMAKE_HAVE_LIBC_PTHREAD failed with the following output:
Change Dir: G:/SourceCode/onnxruntime-1.6.0/onnxruntime-1.6.0/build/Windows/RelWithDebInfo/CMakeFiles/CMakeTmp

Run Build Command(s):C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/MSBuild/Current/Bin/MSBuild.exe cmTC_e1e6d.vcxproj /p:Configuration=Debug /p:Platform=x64 /p:VisualStudioVersion=16.0 /v:m && 用于 .NET Framework 的 Microsoft (R) 生成引擎版本 16.9.0+5e4b48a27

版权所有(C) Microsoft Corporation。保留所有权利。



  用于 x64 的 Microsoft (R) C/C++ 优化编译器 19.28.29914 版

  版权所有(C) Microsoft Corporation。保留所有权利。

  src.c

  cl /c /I"C:\software\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\include" /Zi /W1 /WX- /diagnostics:column /Od /Ob0 /D WIN32 /D _WINDOWS /D CMAKE_HAVE_LIBC_PTHREAD /D "CMAKE_INTDIR=\"Debug\"" /D _MBCS /Gm- /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /Fo"cmTC_e1e6d.dir\Debug\\" /Fd"cmTC_e1e6d.dir\Debug\vc142.pdb" /Gd /TC /errorReport:queue "G:\SourceCode\onnxruntime-1.6.0\onnxruntime-1.6.0\build\Windows\RelWithDebInfo\CMakeFiles\CMakeTmp\src.c"

src.obj : error LNK2019: 无法解析的外部符号 __imp_pthread_create，函数 main 中引用了该符号 [G:\SourceCode\onnxruntime-1.6.0\onnxruntime-1.6.0\build\Windows\RelWithDebInfo\CMakeFiles\CMakeTmp\cmTC_e1e6d.vcxproj]

src.obj : error LNK2019: 无法解析的外部符号 __imp_pthread_detach，函数 main 中引用了该符号 [G:\SourceCode\onnxruntime-1.6.0\onnxruntime-1.6.0\build\Windows\RelWithDebInfo\CMakeFiles\CMakeTmp\cmTC_e1e6d.vcxproj]

src.obj : error LNK2019: 无法解析的外部符号 __imp_pthread_exit，函数 main 中引用了该符号 [G:\SourceCode\onnxruntime-1.6.0\onnxruntime-1.6.0\build\Windows\RelWithDebInfo\CMakeFiles\CMakeTmp\cmTC_e1e6d.vcxproj]

src.obj : error LNK2019: 无法解析的外部符号 __imp_pthread_join，函数 main 中引用了该符号 [G:\SourceCode\onnxruntime-1.6.0\onnxruntime-1.6.0\build\Windows\RelWithDebInfo\CMakeFiles\CMakeTmp`

****************************************************************************************
**PS:**
I found that it is impossible to link to pthread, but I have put include and lib in the vs2019 directory ("C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\ ..."); the 64-bit pthreadVC2.dll is placed in "C:\Windows\System32", and the 32-bit pthreadVC2.dll is placed in "C:\Windows\SysWOW64"
Both x86 and x64 compiled by me can run normally in my test project, and this problem does not occur.
How can I make it load pthreads correctly?

**Reply by snnn** [reply]
On Windows, you don't need pthread.

**Reply by snnn** [reply]
Is the folder  G:/SourceCode/onnxruntime-1.6.0/onnxruntime-1.6.0/cmake/external/googletest empty?

**Reply by shuiyu486** [reply]
Yes it is. G:\SourceCode\onnxruntime-1.6.0\onnxruntime-1.6.0\cmake\external\googletest this folder is empty

**Reply by shuiyu486** [reply]
This is the source code downloaded from the official GitHub, not downloaded using git, the link is: https://github.com/microsoft/onnxruntime/archive/refs/tags/v1.6.0.zip

**Reply by snnn** [reply]
Please download through git. The zip file was auto-generated and uploaded by github, not us. The zip file misses important files.

**Reply by shuiyu486** [reply]
alright, thank you very much

**Reply by shuiyu486** [reply]
Thanks again for your help. After these problems, I successfully built, and I also built the cuda version.
But in the process of using, I found two problems:
1.git clone --recursive https://github.com/Microsoft/onnxruntime v1.6.0 The downloaded version is still 1.7.0
![image](https://user-images.githubusercontent.com/38516221/115545022-e6de6380-a2d5-11eb-8942-8a706a6309ee.png)

2. The cuda version I built will print out the data when the session is run for the first time (this->m_Session->Run). I don't know how it appears (see the picture below). How can I disable it? (Ps: These will not be printed when OrtSessionOptionsAppendExecutionProvider_CPU uses cpu calculations)
![image](https://user-images.githubusercontent.com/38516221/115545606-961b3a80-a2d6-11eb-9074-12190fdd4f0e.png)

This is my instruction to build the cuda version:
.\build.bat --config Release --build_shared_lib --skip_tests --use_openmp --parallel --build_wheel --use_cuda --cuda_version=10.2 --cudnn_home "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2" --cuda_home "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2" --cmake_generator "Visual Studio 16 2019"

Thanks again for your help！

**Reply by snnn** [reply]
@shuiyu486 , when you do "git clone", you will get a whole copy of the repo with all the versions.  So please remove "v1.6.0" from your git command. 
When the clone is completed, you can use commands like:
```
git checkout v1.6.0
```
switching between versions.

**Reply by shuiyu486** [reply]
It turned out to be so, thanks. Do you have any thoughts on Question 2?

**Comment by liqunfu** [comment]
remove the build folder after 'git submodule update --init --recursive' and rebuild

**Reply by shuiyu486** [reply]
> remove the build folder after 'git submodule update --init --recursive' and rebuild
> 

Thanks for your reply, but I have tried this method, and the problem in https://github.com/microsoft/onnxruntime/discussions/7382#discussioncomment-636496 still appears, I don’t know what to do, please help me

**Comment by snnn** [comment]
You didn't install git. 

Please get it from: https://git-scm.com/download/win

Then run the build command as usual.

**Reply by shuiyu486** [reply]
Thank you for your reply. At the beginning, I did not install git. After I installed git, the problem in https://github.com/microsoft/onnxruntime/discussions/7382#discussioncomment-636496 appeared, I don’t know. What to do, please help me

**Reply by shuiyu486** [reply]
![image](https://user-images.githubusercontent.com/38516221/115483097-b3261e00-a282-11eb-969e-190dcb3afcb6.png)
